File: solution_task.py
class Product:

    def __init__(self, uid: str, name: str):
        self.uid = uid
        self.name = name


class ProductsImpl:

    def __init__(self):
        self.products = []

    def add_product(self, product: Product) -> bool:
        for p in self.products:
            if p.uid == product.uid:
                return False
        self.products.append(product)
        return True

    def delete_product(self, product: Product) -> bool:
        for p in self.products:
            if p.uid == product.uid:
                self.products.remove(p)
                return True
        return False

    def get_name(self, uid: str) -> str:
        for product in self.products:
            if product.uid == uid:
                return product.name
        return ""

    def find_by_name(self, name: str) -> [str]:
        result = [product.uid for product in self.products if product.name == name]
        return result

File: solution_task.swift
struct Product {
    let uid: String
    let name: String
}

class ProductsImpl {
    private var products: [Product] = []


    func addProduct(_ product: Product) -> Bool {
        for p in products {
            if p.uid == product.uid {
                return false
            }
        }
        products.append(product)
        return true
    }

    func deleteProduct(_ product: Product) -> Bool {
        if let index = products.firstIndex(where: { $0.uid == product.uid }) {
            products.remove(at: index)
            return true
        }
        return false
    }

    func getName(_ uid: String) -> String {
        if let product = products.first(where: { $0.uid == uid }) {
            return product.name
        }
        return ""
    }

    func findByName(_ name: String) -> [String] {
        let result = products.filter { $0.name == name }.map { $0.uid }
        return result
    }
}
